### PR TITLE
feat(sfn): support ItemReader InputType CSV

### PR DESCRIPTION
**Pull request by ovi-fabric**
## Summary

A Distributed Map can now read a `CSV` dataset through `ItemReader`. Field names come from the first row, or from `ReaderConfig.CSVHeaders` when `CSVHeaderLocation` is `GIVEN`, and each data row becomes an object of string values. `ReaderConfig.CSVDelimiter` selects `COMMA`, `PIPE`, `SEMICOLON`, `SPACE` or `TAB`.

Parsing reuses `CsvParser`, the shared helper Athena, S3 Select and API Gateway also use, and now follows the escaping rules AWS documents for text delimited files:

- a quoted field may carry the delimiter or a line break,
- a doubled quote stands for a single quote, in an unquoted field as well as a quoted one,
- a backslash escapes another backslash, a quote or the delimiter, and a backslash before anything else is dropped.

`parseLine` and `parseAll` share one scanner, so those rules live in a single place.

This followed #3602 (`JSONL`), which has since merged, so the diff here is the CSV work alone.

## Type of change

- [ ] Bug fix (`fix:`)
- [x] New feature (`feat:`)
- [ ] Breaking change (`feat!:` or `fix!:`)
- [ ] Docs / chore

## AWS Compatibility

The parser is pinned against the file and state machine from the [reference example](https://docs.aws.amazon.com/step-functions/latest/dg/example-csv-parse-dist-map.html), including the unquoted `{""MyKey"":""MyValue""}` field that only survives `States.StringToJson` if its doubled quotes are preserved. A row shorter than the headers pads with empty strings, a longer one drops the surplus, and every value is a string.

Backslash handling stays off for `parseLine`, the entry point S3 Select and Athena use. Those services read ordinary CSV, where a lone backslash is data: switching the rules on globally would turn an Athena `C:\Users\x` into `C:Usersx`. Happy to revisit if you would rather the two paths behave the same.

An unsupported `CSVHeaderLocation` or `CSVDelimiter` fails the state at runtime with `States.ItemReaderFailed` rather than being rejected when the state machine is created, which is where AWS rejects it. `validateItemReader` does not inspect these fields today; adding that seemed like a separate change; say the word and I will fold it in.

## Checklist

- [x] `./mvnw test` passes locally (targeted suites, see below)
- [x] New or updated integration test added
- [x] Commit messages follow [Conventional Commits](https://www.conventionalcommits.org/)

`CsvParserTest` is new and covers the reference record, doubled quotes in both quoted and unquoted fields, the three backslash escapes, a dropped backslash, a non comma delimiter, and the `parseLine` contract. Seven ItemReader integration tests cover headers from the first row, `GIVEN` headers, short row padding, a non comma delimiter, a newline inside a quoted field, the full AWS reference file asserted through `States.StringToJson`, and backslash escaping.

```
./mvnw test -Dtest='StepFunctions*,CsvParserTest,ApiKeyCsvParserTest,ApiGatewayImportApiKeysIntegrationTest,S3Select*Test,Athena*Test'
Tests run: 608, Failures: 0, Errors: 0, Skipped: 0
```

The other `CsvParser` consumers are in that run because the parser is shared. The full suite was not run locally because it exhausts memory on this machine, so CI is the first full-suite signal. `make docs-check` passes.

### CODE REVIEW

File: src/main/java/io/github/hectorvent/floci/services/stepfunctions/AslExecutor.java [hunk 1]
```diff
@@ -2868,6 +2877,57 @@ private ResolvedMapItems resolveItemReaderItems(JsonNode itemReader, JsonNode in
...
+    }
+
+    /**
+     * Each data row becomes an object keyed by the headers. A row shorter than the headers pads
+     * with empty strings and a longer one drops the surplus, as on AWS. Every value is a string.
+     */
+    private ArrayNode readCsvRows(JsonNode itemReader, byte[] data) {
+        JsonNode readerConfig = itemReader.path("ReaderConfig");
+        String headerLocation = readerConfig.path("CSVHeaderLocation").asText("FIRST_ROW");
+        List<List<String>> rows = CsvParser.parseAll(new String(data, StandardCharsets.UTF_8));
```
Comment (KenkoGeek) [inline src/main/java/io/github/hectorvent/floci/services/stepfunctions/AslExecutor.java]: The shared parser is not enough for AWS ItemReader CSV semantics yet. AWS documents both `CSVDelimiter` (COMMA/PIPE/SEMICOLON/SPACE/TAB) and quoted fields that may contain newlines without ending the record. This call always parses comma-delimited text and `CsvParser.parseAll` splits on line boundaries first, so valid AWS datasets such as `CSVDelimiter: PIPE` or a quoted multi-line field are misparsed. Since this PR is adding `InputType: CSV`, I would treat those as part of the feature contract rather than a later enhancement. Please honor `ReaderConfig.CSVDelimiter` and preserve newlines inside quoted fields, with focused regressions for both. AWS: https://docs.aws.amazon.com/step-functions/latest/dg/input-output-itemreader.html

**Review by KenkoGeek — CHANGES_REQUESTED** [review]
This is a strong addition overall, and I appreciate that you called out the parser limitation explicitly instead of hiding it. The header handling, ragged-row behavior, string typing, and stacked-PR description are all thoughtfully done. I found one compatibility gap in the core CSV parsing contract that should be closed before merge.

**Comment by ovi-fabric** [comment]
Thanks, both are now handled. `CsvParser.parseAll` takes a delimiter and scans records character by character, so a line break inside a quoted field no longer ends the record, and `ReaderConfig.CSVDelimiter` maps COMMA, PIPE, SEMICOLON, SPACE and TAB, with anything else failing as `States.ItemReaderFailed`. Added regressions for a PIPE dataset and a quoted multi-line field; the CSV tests pass locally (5/5). Docs updated.

**Review by KenkoGeek — COMMENTED** [review]
Great follow-up, thanks for closing both parts of the CSV contract rather than only adding the delimiter switch. The new parser now keeps quoted line breaks inside the same record, `CSVDelimiter` covers the documented delimiter values, and the focused PIPE plus multiline regressions exercise both behaviors directly. My CSV compatibility finding is resolved on this head.

I am not approving yet only because required CI is still red on the unrelated CloudFormation rollback test. There is no remaining CSV finding from my side.

**Comment by ovi-fabric** [comment]
Correction on the CI red here: it was not the CloudFormation flake I assumed. `distributedMapWithCsvItemReader_failsWithNotImplementedItemReaderError` asserts that `InputType: CSV` fails as unimplemented, which this branch implements, so it had been failing since the CSV commit. I had only run the focused CSV tests on this branch and missed it. Removed that assertion, since the CSV reader tests now cover the behaviour, and merged latest main. Full `StepFunctionsJsonataIntegrationTest` passes locally (62/62).

**Comment by ovi-fabric** [comment]
CI is green on this head now (24/24), after removing the stale CSV not-implemented assertion and merging main.

**Review by KenkoGeek — APPROVED** [review]
Great follow-through on the CSV reader. Thanks for catching that the remaining red was actually a stale not-implemented assertion on this branch rather than an unrelated flake, and for removing the obsolete expectation instead of masking it. The CSV delimiter and multiline-record behavior were already correct from the previous pass; this head now has the full required matrix green and is CLEAN/MERGEABLE. No remaining Step Functions compatibility blocker from my side.

File: src/main/java/io/github/hectorvent/floci/services/stepfunctions/AslExecutor.java (hunk 1 above)
Comment (ovi-fabric) [reply src/main/java/io/github/hectorvent/floci/services/stepfunctions/AslExecutor.java]: Done in 8e4c5a7-range commit "feat(sfn): honour CSVDelimiter and quoted newlines in ItemReader CSV": CsvParser.parseAll now takes a delimiter and scans records character by character, so a quoted line break no longer ends a record, and CSVDelimiter maps COMMA, PIPE, SEMICOLON, SPACE and TAB. Regressions added for a PIPE dataset and a quoted multi-line field.

**Review by KenkoGeek — APPROVED** [review]
The current head is fully green and CLEAN/MERGEABLE. I rechecked the actual PR patch rather than the noisy merge history: the CSV ItemReader behavior previously reviewed, including delimiter and quoted-newline handling, is unchanged; the newer head only incorporates main/JSONL branch updates. Approved on the current head.

**Review by pgermosen — COMMENTED** [review]
Solid work carrying this through the earlier rounds, but there's a real gap that never got exercised: AWS's own reference example for CSV parsing has an unquoted field like `{""MyKey"":""MyValue""}` collapsing to `{"MyKey":"MyValue"}`, and running that exact input through this parser drops the quotes entirely instead, so `States.StringToJson` on it would produce garbage rather than the object AWS returns. Backslash escaping is documented too (a backslash escapes another backslash, a quote, or the delimiter) and isn't handled anywhere here either. Both live in the same quote-tracking loop, which is now duplicated between parseLine and the new parseAll rather than shared, so whatever fix lands for one has to be applied twice. Everything the previous rounds focused on, quoted newlines and commas, delimiter selection, short/long row padding, held up fine against the docs on my pass.

**Review by KenkoGeek — CHANGES_REQUESTED** [review]
Re-reviewed the new head after the latest main merge. Priamo's CSV escaping finding is valid and still open: AWS explicitly requires doubled quotes to preserve quotes and backslash escaping for backslash, quote, and the configured delimiter. Those semantics matter directly for documented `States.StringToJson` CSV examples, so this is AWS-visible behavior rather than a parser nicety. My prior approval predates this head and should not carry forward. Please cover the documented quote/backslash cases in the shared CsvParser and ItemReader regression tests before merge.

**Comment by pgermosen** [comment]
There's a new repo rule that blocks this regardless of the code review: main now requires every commit to be signed, and none of the 13 commits on this branch are. GitHub will refuse the merge outright once that's enforced, independent of review approval or CI, so this needs sorting before it can land.

Could you turn on commit signing (GPG or SSH, whichever you've got set up with GitHub) and re-push? An interactive rebase with signing enabled, then a force-push, should cover all of them in one pass.

**Comment by ovi-fabric** [comment]
Both findings confirmed and fixed in dc2ff063e. Thanks, the reference example was the right thing to point at.

**Doubled quotes.** Reproduced exactly as described: `{""MyKey"":""MyValue""}` came out as `{MyKey:MyValue}`, so `States.StringToJson` had nothing valid to parse. The cause was narrower than it looked. Only a quote that *opens* a field should start a quoted run; the loop let any quote toggle, so in an unquoted field the `""` pair opened a run and immediately closed it, consuming both characters. Restricting the toggle to a field-opening quote makes every documented case fall out, and it leaves `""` as a whole field an empty string, which is what the other callers expect of ordinary CSV.

**Backslashes.** Also missing, and worse than reported: `a\"b` yielded `a\b` and swallowed the following newline, because the `\"` opened a quoted run that then ate the record separator. Now a backslash escapes another backslash, a quote or the configured delimiter, and a backslash before anything else is dropped, per the docs.

**Shared loop.** `parseLine` and `parseAll` now run off one scanner, so there is a single place to change.

One deliberate asymmetry worth your call: the backslash rules are **off** for `parseLine`, which S3 Select and Athena use. Those read ordinary CSV where a lone backslash is data, so enabling the rules there would turn an Athena `C:\Users\x` into `C:Usersx`. The quote fix is shared, since it is correct for both. Say the word if you would rather the two paths behave identically.

**Coverage.** `CsvParserTest` is new (the shared parser had no unit test): the reference record, doubled quotes in quoted and unquoted fields, the three backslash escapes, a dropped backslash, a non comma delimiter, and the `parseLine` contract. Two ItemReader integration tests were added, one being the AWS reference file with the reference state machine, asserted through `States.StringToJson` so `MyObject.MyKey == "MyValue"`.

```
./mvnw test -Dtest='StepFunctions*,CsvParserTest,ApiKeyCsvParserTest,ApiGatewayImportApiKeysIntegrationTest,S3Select*Test,Athena*Test'
Tests run: 608, Failures: 0, Errors: 0, Skipped: 0
```

The branch was also rebuilt on current `main` as three signed commits, so the earlier merge commits and the duplicated `JSONL` commits from #3602 are gone.

**Review by pgermosen — APPROVED** [review]
Traced the reference field through the fixed scanner by hand rather than take the write-up at face value, and it produces exactly what AWS documents, restricting the quote-toggle to a real field-opening quote is the right fix. Also checked the shared-scanner refactor doesn't regress the other CsvParser consumers, since it changes parseLine's quote handling too, their suites are all green and there's a test for the quoted-empty-field edge case that could've gone wrong. Approving.

**Review by pgermosen — COMMENTED** [review]
All three things from last round landed and I traced them by hand against AWS's own reference example, doubled quotes in and out of a quoted field, all three backslash escapes, they match exactly, and the new integration test runs the actual AWS reference CSV end to end through StartExecution and checks the documented output verbatim, which is about as good a parity check as this gets without a real account.

One loose thread: unifying parseLine onto the shared scanner also changes what it does with a stray unquoted `"` that isn't part of a doubled pair, and neither the new tests nor anything existing seems to exercise that for Athena or S3 Select. Probably fine, likely better than the old behavior even, just worth a quick look since those are different engines on real AWS.
